### PR TITLE
[ONNX] export group_norm

### PR DESCRIPTION
**Pull request by neginraoof**
Updated group_norm symbolic

### CODE REVIEW

File: test/onnx/test_operators.py [hunk 1]
```diff
@@ -563,6 +563,15 @@ def test_unsqueeze(self):
...
         self.assertONNX(lambda x: x.unsqueeze(len(x.shape)), x)
 
+    def test_groupnorm(self):
+        x = torch.randn(4, 6, 10, 10, 10, requires_grad=True)
+        self.assertONNX(nn.GroupNorm(3, 6, eps=0.002, affine=False), x)
+
+    def test_groupnorm_noaffine(self):
+        model = torch.nn.GroupNorm(3, 6, 0.4)
+        x = torch.randn(3, 8, 10, 10, requires_grad=True)
+        self.assertONNX(nn.GroupNorm(4, 8, eps=0.002), x)
```
Comment (BowenBao) [inline test/onnx/test_operators.py]: why is this test case called 'noaffine', but affine is True? (vs the above other test case where affine=False).

File: torch/onnx/symbolic_opset9.py [hunk 2]
```diff
@@ -1950,15 +1944,47 @@ def multinomial(g, input, num_samples, replacement=False, generator=None):
...
 def gelu(g, self):
     _sqrt2 = 1.4142135623730951
     erf = g.op('Erf', div(g, self, torch.tensor(_sqrt2)))
     erf_plusone = add(g, erf, g.op('Constant', value_t=torch.tensor(1, dtype=torch.float)))
     return mul(g, mul(g, self, erf_plusone), g.op('Constant', value_t=torch.tensor(0.5, dtype=torch.float)))
+
+
+@parse_args('v', 'i', 'v', 'v', 'f', 'i')
+def group_norm(g, input, num_groups, weight, bias, eps, cudnn_enabled):
+    input_sizes = input.type().sizes()
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py]: This requires input shape to be known. Consider adding a helper function in symbolic_helper.py, since this pattern appears in a lot of places.

File: torch/onnx/symbolic_opset9.py [hunk 3]
```diff
@@ -1950,15 +1944,47 @@ def multinomial(g, input, num_samples, replacement=False, generator=None):
...
+    shape = [1, input_sizes[0] * num_groups, input_sizes[1] / num_groups, -1]
+    input_reshaped = g.op('Reshape', input, g.op('Constant', value_t=torch.LongTensor(shape)))
+
+    # Due to shape difference. we need to apply weight and bias after
+    # instance norm computation and reshape
+    weight_ = g.op("Constant", value_t=torch.tensor([1.0] * input_sizes[0] * num_groups))
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py]: Maybe the ` * input_sizes[0] * num_groups` is not necessary, if broadcasting is supported.

File: torch/onnx/symbolic_opset9.py [hunk 4]
```diff
@@ -1950,15 +1944,47 @@ def multinomial(g, input, num_samples, replacement=False, generator=None):
...
+    bias_ = g.op("Constant", value_t=torch.tensor([0.0] * input_sizes[0] * num_groups))
+
+    norm_reshaped = g.op("InstanceNormalization", input_reshaped, weight_, bias_, epsilon_f=eps)
+    norm = g.op('Reshape', norm_reshaped, g.op('Constant', value_t=torch.LongTensor(input_sizes)))
+
+    if weight is None or weight.node().mustBeNone():
+        assert len(input_sizes) > 1
+        weight_value = torch.tensor([1.] * input_sizes[1]).type(
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py]: Same as above, check if we can remove the `* input_sizes[1]` part. Fewer dependencies the better.

File: torch/onnx/symbolic_opset9.py [hunk 5]
```diff
@@ -1950,15 +1944,47 @@ def multinomial(g, input, num_samples, replacement=False, generator=None):
...
+@parse_args('v', 'i', 'v', 'v', 'f', 'i')
+def group_norm(g, input, num_groups, weight, bias, eps, cudnn_enabled):
+    input_sizes = input.type().sizes()
+    shape = [1, input_sizes[0] * num_groups, input_sizes[1] / num_groups, -1]
+    input_reshaped = g.op('Reshape', input, g.op('Constant', value_t=torch.LongTensor(shape)))
+
+    # Due to shape difference. we need to apply weight and bias after
+    # instance norm computation and reshape
+    weight_ = g.op("Constant", value_t=torch.tensor([1.0] * input_sizes[0] * num_groups))
+    bias_ = g.op("Constant", value_t=torch.tensor([0.0] * input_sizes[0] * num_groups))
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py]: Here the data type of value_t for `weight_` and `bias_` are always float32, is this ok for cases where input is Double?

File: torch/onnx/symbolic_opset9.py (hunk 2 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py]: I'm adding a check for if input.isCompleteTensor()
I can add a helper function which does this check internally, but just returning input sizes is not that helpful

File: torch/onnx/symbolic_opset9.py [hunk 6]
```diff
@@ -1950,15 +1944,47 @@ def multinomial(g, input, num_samples, replacement=False, generator=None):
...
     _sqrt2 = 1.4142135623730951
     erf = g.op('Erf', div(g, self, torch.tensor(_sqrt2)))
     erf_plusone = add(g, erf, g.op('Constant', value_t=torch.tensor(1, dtype=torch.float)))
     return mul(g, mul(g, self, erf_plusone), g.op('Constant', value_t=torch.tensor(0.5, dtype=torch.float)))
+
+
+@parse_args('v', 'i', 'v', 'v', 'f', 'i')
+def group_norm(g, input, num_groups, weight, bias, eps, cudnn_enabled):
+    input_sizes = input.type().sizes()
+    shape = [1, input_sizes[0] * num_groups, input_sizes[1] / num_groups, -1]
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py]: In fact it might be possible to remove the dependency on explicit `input_sizes`. 

- For the first reshape, reshape to [0, num_groups, -1], the effect will be [batch, channel, h, w] => [batch, num_groups, channel / num_groups * h * w].
- For `weight` and `bias`, consider the below comment regarding broadcasting.
- For the second reshape, just reshape back to the original input shape: `g.op('Reshape', norm_reshaped, g.op("Shape", input))`.
- For the last `weight` and `bias`, it seems we still need the rank information of input. We can remove this dependency as well if we add two more reshapes to the norm tensor, but that might not worth the effort as rank info is usually static. 

The motivation is that we won't be able to support dynamic input sizes if we depend on static `input_sizes`.

File: torch/onnx/symbolic_opset9.py (hunk 3 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py]: I checked this. If I don't create weight of correct shape, I'll get ORT error:
size of scale != input channel count 1 vs. 12

File: torch/onnx/symbolic_opset9.py (hunk 2 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py]: I think I can send this out as a new PR, don't really want to touch other ops in this PS

File: torch/onnx/symbolic_opset9.py (hunk 4 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py]: Yes, this is not needed

File: torch/onnx/symbolic_opset9.py (hunk 5 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py]: This is fixed in new commit.

File: torch/onnx/symbolic_opset9.py (hunk 6 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py]: Thanks @BowenBao, apparently broadcasting is not supported for InstanceNormalization.
So I don't think we can avoid accessing the size for weight_ and bias_

Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py]: Ok, but I think we can still construct weight_ and bias_ with just num_groups, if in the first reshape we keep batch and channel axes separated.

Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py]: True, I gonna update the first two parts anyways. But I agree that it might not be worth it to add two reshapes since rank here could be N (it's not limited in spec). I'll leave a note in case anyone needed this op with dynamic rank in future.

File: torch/onnx/symbolic_opset9.py [hunk 7]
```diff
@@ -1950,15 +1944,50 @@ def multinomial(g, input, num_samples, replacement=False, generator=None):
...
 def gelu(g, self):
     _sqrt2 = 1.4142135623730951
     erf = g.op('Erf', div(g, self, torch.tensor(_sqrt2)))
     erf_plusone = add(g, erf, g.op('Constant', value_t=torch.tensor(1, dtype=torch.float)))
     return mul(g, mul(g, self, erf_plusone), g.op('Constant', value_t=torch.tensor(0.5, dtype=torch.float)))
+
+
+@parse_args('v', 'i', 'v', 'v', 'f', 'i')
+def group_norm(g, input, num_groups, weight, bias, eps, cudnn_enabled):
+    if input.isCompleteTensor():
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py]: isCompleteTensor can be relaxed. We only require its scalar type and rank to be known.

File: test/onnx/test_pytorch_onnx_caffe2.py [hunk 8]
```diff
@@ -1473,9 +1473,16 @@ def forward(self, x):
         x = torch.randn(16, 3, 256, 256)
         self.run_model_test(ReduceSumMultipleAxes(), train=False, input=(x,), batch_size=BATCH_SIZE, use_gpu=False)
 
+    @skipIfEmbed
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_caffe2.py]: why is this needed?

**Review by BowenBao — COMMENTED** [review]
please always remember to update test operator expect files if symbolic fn is updated.

File: test/onnx/test_pytorch_onnx_caffe2.py (hunk 8 above)
Comment (neginraoof) [reply test/onnx/test_pytorch_onnx_caffe2.py]: https://github.com/pytorch/pytorch/blob/7e2a5c34e31cbbc932a59d47c9860b42f88672dc/test/onnx/test_pytorch_onnx_caffe2.py#L1231
Apparently instance norm includes unused weights.

File: torch/onnx/symbolic_opset9.py (hunk 7 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py]: Sure. I'm on console access to the environment and build is slow, so I updated the files here.

**Comment by neginraoof** [comment]
cc @houseroad tests are fixed. Please re-import. Thanks.

**Comment by neginraoof** [comment]
cc @houseroad for review

**Comment by neginraoof** [comment]
cc @houseroad for review

File: torch/onnx/symbolic_opset9.py [hunk 9]
```diff
@@ -1970,3 +1966,33 @@ def gelu(g, self):
     erf = g.op('Erf', div(g, self, torch.tensor(_sqrt2)))
     erf_plusone = add(g, erf, g.op('Constant', value_t=torch.tensor(1, dtype=torch.float)))
     return mul(g, mul(g, self, erf_plusone), g.op('Constant', value_t=torch.tensor(0.5, dtype=torch.float)))
+
+
+@parse_args('v', 'i', 'v', 'v', 'f', 'i')
+def group_norm(g, input, num_groups, weight, bias, eps, cudnn_enabled):
+    input_sizes = input.type().sizes()
+    shape = [0, num_groups, -1]
```
Comment (houseroad) [inline torch/onnx/symbolic_opset9.py]: what's the meaning of 0?

This reshape doesn't look good to me.

File: torch/onnx/symbolic_opset9.py [hunk 10]
```diff
@@ -1970,3 +1966,33 @@ def gelu(g, self):
...
+    input_reshaped = g.op('Reshape', input, g.op('Constant', value_t=torch.LongTensor(shape)))
+
+    # Due to shape difference. we need to apply weight and bias after
```
Comment (houseroad) [inline torch/onnx/symbolic_opset9.py]: Is instance norm can handle all the cases? If dim C is not divisible by group, are we still able to use instancenorm to represent groupnorm?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 11]
```diff
@@ -498,6 +498,16 @@ def test_interpolate_upsample(self):
     def test_interpolate_downsample(self):
         self._interpolate_tests(False)
 
+    def test_groupnorm(self):
```
Comment (houseroad) [inline test/onnx/test_pytorch_onnx_onnxruntime.py]: More thorough end to end tests please

**Review by houseroad — CHANGES_REQUESTED** [review]
Please check the inline comments

File: torch/onnx/symbolic_opset9.py (hunk 10 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py]: What I understanf from the spec (https://pytorch.org/docs/stable/_modules/torch/nn/modules/normalization.html#GroupNorm)
is that C is always divisible by group

Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py]: If this is not the case, torch complains, for example:
RuntimeError: Expected number of channels in input to be divisible by num_groups, but got input of shape [20, 6, 10, 10] and num_groups=4

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 11 above)
Comment (neginraoof) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: Are you asking for different channel/num group cases?

File: torch/onnx/symbolic_opset9.py (hunk 9 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py]: onnx::reshape docs mention that using 0 in shape tensor let's you to infer the shape from input.
https://github.com/onnx/onnx/blob/master/docs/Operators.md#reshape
(actual dimension value would be unchanged). We are using this to avoid statically accessing the shape.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 11 above)
Comment (neginraoof) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: I addressed the two comments about the logic and commented in the code. Please let me know
which part is missing or not covered in the tests.

**Comment by neginraoof** [comment]
@houseroad please review updates.

**Comment by neginraoof** [comment]
cc @houseroad for review

**Comment by neginraoof** [comment]
@houseroad for review

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 11 above)
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: You can add the below additional test cases https://pytorch.org/docs/master/nn.html#groupnorm:

> \# Separate 6 channels into 6 groups (equivalent with InstanceNorm)
> m = nn.GroupNorm(6, 6)
> \# Put all 6 channels into a single group (equivalent with LayerNorm)
> m = nn.GroupNorm(1, 6)

To test https://github.com/pytorch/pytorch/blob/aea140523a5c7fdf017439fbcc461cafd95c66d4/torch/onnx/symbolic_opset9.py#L1987
try construct test case with `weight`, `bias` being None

Comment (neginraoof) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: Thanks @BowenBao. I added the tests for various num_groups.
I just verified case where weight/bias is none is covered in test_groupnorm_noaffine test.

File: test/onnx/test_operators.py [hunk 12]
```diff
@@ -563,6 +563,14 @@ def test_unsqueeze(self):
         x = torch.randn(3, 4, requires_grad=True)
         self.assertONNX(lambda x: x.unsqueeze(len(x.shape)), x)
 
+    def test_groupnorm(self):
```
[the remark this answers is not in the document]
Comment (neginraoof) [reply test/onnx/test_operators.py]: Sure. Removed this test

**Comment by neginraoof** [comment]
@houseroad please review the updates

**Review by houseroad — CHANGES_REQUESTED** [review]
This broke some internal use cases, I need further investigation.

**Comment by neginraoof** [comment]
Thanks @houseroad for looking into this. Is there a failure regarding group_norm?

**Review by houseroad — APPROVED** [review]
Let's check in this for now. However, this translation is really inefficient. We should introduce GroupNorm to ONNX, so we can use one op to represent the model.
